lib/routes/account.js

```diff
@@ -1612,4 +1630,50 @@ module.exports = function (
   }
 
   return routes
+
+  function upsertDevice (request, sessionToken, deviceInfo) {
+    var operation = deviceInfo.id ? 'updateDevice' : 'createDevice'
+    return db[operation](sessionToken.uid, sessionToken.tokenId, deviceInfo)
+      .then(
+        function (device) {
+          if (operation === 'createDevice') {
+            log.event('device:create', request, {
+              uid: sessionToken.uid,
+              id: device.id,
+              type: device.type,
+              timestamp: device.createdAt
+            })
+          }
+          push.notifyDeviceConnected(sessionToken.uid, device.name, device.id.toString('hex'))
+          return device
+        }
+      )
+  }
+
+  function synthesizeDeviceName (device) {
```

Comment:
Although this function is only called from the `/account/devices` endpoint, I've thrown it down here next to `upsertDevice` because they're conceptually related. Maybe there is a case for adding a new `lib/devices` module?